src/build.d

```diff
@@ -742,6 +801,8 @@ auto sourceFiles()
         lexer:
             lexerDmdFiles.map!(e => env["D"].buildPath(e ~ ".d")).chain(
             lexerRootFiles.map!(e => env["ROOT"].buildPath(e ~ ".d"))).array,
+        parser: "parse astbase parsetimevisitor transitivevisitor permissivevisitor strictvisitor utils"
+            .split.map!(e => env["D"].buildPath(e ~ ".d")).array,
```

Comment:
Why can't we define the sources as an array directly?

Reply:
Not needing the extra 4 characters per file `foo", "bar` seems to save a good amount of space and makes it easier to look at and modify the lists.  What's the reason for using an array?  Are we concerned with performance?

Reply:
An array is an order of magnitude easier to read than this.

Reply:
An order of magnitude harder to read? How do you calculate the reading difficulty?

Reply:
So you're saying this this is easier to read:
```
["a", "b", "c", "d", "e", "f", "g", "h" "i"]
```
Than this?
```
"a b c d e f g h i"
```
What's the reasoning?

Reply:
No, I'm saying it's an order of magnitude easier to read:
```d
parser: [
    "parse.d",
    "astbase.d",
    "parsetimevisitor.d", "transitivevisitor.d", "permissivevisitor.d", "strictvisitor.d"
    "utils.d",
],
```

Than what's currently there, because it does not require me to do any computation to understand what the data actually is.

Reply:
Ok I can see that argument.  I wouldn't equate that to an order of magnitude but you're free to your opinion.  However, in this case it's not me who necessarily disagrees with you, it's whoever wrote the makefiles.  In the makefiles they list the files without extensions and add the extensions through post-processing just like I'm doing here.

Reply:
And your example could also be written as this:
```D
parser: "
    parse.d
    astbase.d
    parsetimevisitor.d transitivevisitor.d permissivevisitor.d strictvisitor.d
    utils.d
"
```
I'm still not seeing how using an array vs splitting a string is easier to read?

Reply:
Isn't the whole point of rewriting the makefiles in D so that they can be idiomatic D code? This is a Makefile/scripting idiom.

Reply:
Operations on strings are the source of many bugs. To give a silly example, the code would currently break for filenames containing spaces or newlines.

Reply:
Does D have an idiom that dictates how to create arrays of strings?  Can you point me to it?

So far we're just talking about what's easier to read though.  So we have this:
```D
"access.d aggregate.d aliasthis.d apply.d argtypes.d argtypes_sysv_x64.d arrayop.d
arraytypes.d astcodegen.d ast_node.d attrib.d builtin.d canthrow.d cli.d clone.d compiler.d complex.d cond.d constfold.d
cppmangle.d cppmanglewin.d ctfeexpr.d ctorflow.d dcast.d dclass.d declaration.d delegatize.d denum.d dimport.d
dinifile.d dinterpret.d dmacro.d dmangle.d dmodule.d doc.d dscope.d dstruct.d dsymbol.d dsymbolsem.d
dtemplate.d dversion.d escape.d expression.d expressionsem.d func.d
hdrgen.d id.d impcnvtab.d imphint.d init.d initsem.d inline.d inlinecost.d intrange.d
json.d lambdacomp.d lib.d libelf.d libmach.d link.d mars.d mtype.d nogc.d nspace.d objc.d opover.d optimize.d parse.d permissivevisitor.d sapply.d templateparamsem.d
sideeffect.d statement.d staticassert.d target.d typesem.d traits.d transitivevisitor.d parsetimevisitor.d visitor.d
typinf.d utils.d scanelf.d scanmach.d statement_rewrite_walker.d statementsem.d staticcond.d safe.d blockexit.d printast.d 
semantic2.d semantic3.d".split
```

vs this:
```D
["access.d", "aggregate.d", "aliasthis.d", "apply.d", "argtypes.d", "argtypes_sysv_x64.d", "arrayop.d",
"arraytypes.d", "astcodegen.d", "ast_node.d", "attrib.d", "builtin.d", "canthrow.d", "cli.d", "clone.d", "compiler.d", "complex.d", "cond.d", "constfold.d",
"cppmangle.d", "cppmanglewin.d", "ctfeexpr.d", "ctorflow.d", "dcast.d", "dclass.d", "declaration.d", "delegatize.d", "denum.d", "dimport.d",
"dinifile.d", "dinterpret.d", "dmacro.d", "dmangle.d", "dmodule.d", "doc.d", "dscope.d", "dstruct.d", "dsymbol.d", "dsymbolsem.d",
"dtemplate.d", "dversion.d", "escape.d", "expression.d", "expressionsem.d", "func.d",
"hdrgen.d", "id.d", "impcnvtab.d", "imphint.d", "init.d", "initsem.d", "inline.d", "inlinecost.d", "intrange.d",
"json.d", "lambdacomp.d", "lib.d", "libelf.d", "libmach.d", "link.d", "mars.d", "mtype.d", "nogc.d", "nspace.d", "objc.d", "opover.d", "optimize.d", "parse.d", "permissivevisitor.d", "sapply.d", "templateparamsem.d",
"sideeffect.d", "statement.d", "staticassert.d", "target.d", "typesem.d", "traits.d", "transitivevisitor.d", "parsetimevisitor.d", "visitor.d
"typinf.d", "utils.d", "scanelf.d", "scanmach.d", "statement_rewrite_walker.d", "statementsem.d", "staticcond.d", "safe.d", "blockexit.d", "printast.d",
"semantic2.d", "semantic3.d"]
```

I'm not saying one is better than the other, but if you ask me which one is easier to read, for it's it's the first example.  However @Geod24 says the second example is an "order of magnitude" easier to read.

Reply:
> Operations on strings are the source of many bugs. To give a silly example, the code would currently break for filenames containing spaces or newlines.

Sure.  Seems like a weak argument in this case though.  I think the "readability" would be a bigger factor here, as I think that would be a bigger source of bugs in this particular case.

Reply:
> Does D have an idiom that dictates how to create arrays of strings? Can you point me to it?

I think the latter is much more common in D code than the former.

> So far we just seem to disagree on what's easier to read. So we have this:
> 
> *snip*
>
> vs this:
> 
> *snip*
> 
> I'm not saying one is better than the other, but if you ask me which one is easier to read then I would say it's the first example. However @Geod24 says the second example is an "order of magnitude" easier to read.

Personally I also find the latter easier to read (i.e. it's a list with individual string literals and I don't have to think about the `split`).

Reply:
> I think the latter is much more common in D code than the former.

Agreed.

> Personally I also find the latter easier to read (i.e. it's a list with individual string literals and I don't have to think about the split).

I agree it's easier to see what it's doing, but have to disagree that it's easier to read.  In any case, not a big deal either way.

Reply:
I think not using UFCS would make it easier to see what is doing, since you see that first. I'm not really fussed either way, though.

Reply:
> Isn't the whole point of rewriting the makefiles in D so that they can be idiomatic D code? This is a Makefile/scripting idiom.

Yes.  Personally, I would rather that the build.d file look nothing like the makefiles and instead be a normal sequential D program, but beggars can't be choosers and I'm just thrilled that others are willing to do some of this work.  Just having this logic in D is already a plus.  It does help when migrating this logic from makefiles for it to resemble the makefiles until the conversion is finished, so I'm willing to let some things slide.  I hope after the conversion is done build.d can evolve to something more idiomatic-D.

Reply:
I'm in a similar boat.  I'm mostly just following the conventions here established by @wilzbach 

At times it can be harder to follow but it makes for smaller code size which I think was important in convincing W&A to make the transition.  And he uses similar techniques throughout so once you get used to it it's easy to maintain.

In the end I prefer larger code size and simpler/more readable code, the but the most important aspect in this case is code size, as this is the most common argument used by Walter and Andrei against accepting changes.  It's also a reason to try not to write too many tests, as they don't distinguish between production/test code, they only care about the line count.